### DISCUSSION TITLE
Rendering without a vertex buffer

### DISCUSSION THREAD

**Discussion by streifmi** [attackgoat/screen-13#41] [Q&A] [answered]
Does screen-13 support rendering without binding a vertex buffer (i.e. only using an index buffer)? I tried the most obvious thing and simply didn't set one during `record_subpass`, but that causes the following validation error:

[2022-10-28T20:47:07Z ERROR screen_13::driver::instance] 🆘 vkCmdDrawIndexed: VkPipeline 0x4fac1c0000000032[] expects that this Command Buffer's vertex binding Index 0 should be set via vkCmdBindVertexBuffers. This is because pVertexBindingDescriptions[0].binding value is 0. The Vulkan spec states: All vertex input bindings accessed via vertex input variables declared in the vertex shader entry point's interface must have either valid or VK_NULL_HANDLE buffers bound (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCmdDrawIndexed-None-04007)

**Comment by attackgoat** [comment, marked as answer]
Thanks for the detailed report!

This is not currently supported, but probably should be. Adding it looks straight forward - I've created #42 to cover this use case. If you feel inclined, please check out the test that change adds to `fuzzer.rs` and see if that works for you. I'd like to improve the API and use the `Into<AnyBufferBinding>` pattern if possible so expect some changes over the next few days.

**Reply by streifmi** [reply]
Thanks for the fast response! I'm a bit surprised that this even requires `VK_NULL_HANDLE`. The validation error might be a bit misleading here. Looking at [this](https://www.saschawillems.de/blog/2016/08/13/vulkan-tutorial-on-rendering-a-fullscreen-quad-without-buffers/) it should be possible to just use an empty vertex input state (and for my use case still bind an index buffer). Though I am by no means a Vulkan expert, so I might be totally wrong of course.

EDIT: I still had vertex attribute bindings in the vertex shader - that's the cause of the validation error. But even if I remove that, it does not work: the screen remains black, but there's no validation errors anymore. I'll investigate further.

**Reply by attackgoat** [reply]
[This code](https://github.com/attackgoat/screen-13/blob/master/contrib/screen-13-fx/res/shader/graphic/present.vert) demonstrates drawing without a vertex buffer _while_ having no vertex input variables in the shader.

I suppose it's kind of an odd thing to declare input in the shader _without_ providing data while drawing. Apparently it's an actual thing people wanted? The PR makes this scenario possible but I'm wondering if it provides any value.
